hbase-common/src/main/java/org/apache/hadoop/hbase/security/oauthbearer/internals/OAuthBearerSaslServer.java

```diff
@@ -0,0 +1,270 @@
+    try {
+      callbackHandler.handle(new Callback[] {callback});
+    } catch (IOException | UnsupportedCallbackException e) {
+      handleCallbackError(e);
+    }
+    OAuthBearerToken token = callback.token();
+    if (token == null) {
+      errorMessage = jsonErrorResponse(callback.errorStatus(), callback.errorScope(),
+        callback.errorOpenIDConfiguration());
+      LOG.error("JWT token validation error: {}", errorMessage);
+      return errorMessage.getBytes(StandardCharsets.UTF_8);
+    }
+    /*
+     * We support the client specifying an authorization ID as per the SASL
+     * specification, but it must match the principal name if it is specified.
+     */
+    if (!authorizationId.isEmpty() && !authorizationId.equals(token.principalName())) {
+      throw new SaslAuthenticationException(String.format(
+        "Authentication failed: Client requested an authorization id (%s) that is different from "
+          + "the token's principal name (%s)",
+        authorizationId, token.principalName()));
+    }
+
+    Map<String, String> validExtensions = processExtensions(token, extensions);
+
+    tokenForNegotiatedProperty = token;
+    this.extensions = new SaslExtensions(validExtensions);
+    complete = true;
+    LOG.debug("Successfully authenticate User={}", token.principalName());
+    return new byte[0];
+  }
+
+  private Map<String, String> processExtensions(OAuthBearerToken token, SaslExtensions extensions)
+    throws SaslException {
+    OAuthBearerExtensionsValidatorCallback
+      extensionsCallback = new OAuthBearerExtensionsValidatorCallback(token, extensions);
+    try {
+      callbackHandler.handle(new Callback[] {extensionsCallback});
+    } catch (UnsupportedCallbackException e) {
+      // backwards compatibility - no extensions will be added
+    } catch (IOException e) {
+      handleCallbackError(e);
+    }
+    if (!extensionsCallback.invalidExtensions().isEmpty()) {
+      String errorMessage = String.format("Authentication failed: %d extensions are invalid! "
+          + "They are: %s", extensionsCallback.invalidExtensions().size(),
+        Utils.mkString(extensionsCallback.invalidExtensions(), "", "", ": ", "; "));
+      LOG.debug(errorMessage);
+      throw new SaslAuthenticationException(errorMessage);
+    }
+
+    return extensionsCallback.validatedExtensions();
+  }
+
+  private static String jsonErrorResponse(String errorStatus, String errorScope,
+    String errorOpenIDConfiguration) {
+    String jsonErrorResponse = String.format("{\"status\":\"%s\"", errorStatus);
+    if (errorScope != null) {
+      jsonErrorResponse = String.format("%s, \"scope\":\"%s\"", jsonErrorResponse, errorScope);
+    }
+    if (errorOpenIDConfiguration != null) {
+      jsonErrorResponse = String.format("%s, \"openid-configuration\":\"%s\"", jsonErrorResponse,
+        errorOpenIDConfiguration);
+    }
+    jsonErrorResponse = String.format("%s}", jsonErrorResponse);
+    return jsonErrorResponse;
```

Comment:
`String.format` is not a good approach to create a JSON. You should take care of the escaping at least.

Reply:
What do you recommend here?
I don't we need full-blown json generating here, because the variables are populated by us, they don't contain user input.